### ISSUE TITLE
Cross-model relations fail after requiring pod changes

### ISSUE THREAD

**Issue by kian99** [canonical/vault-operator#243]
### Bug Description

I have 2 Juju models attached to a single controller. One is a k8s model and the other is a machine model.
In the machine model Vault is deployed.
In the k8s model an app that relates to Vault is deployed.

When I initially relate the two via cross-model relations, there is no issue and `juju show-unit <my-app>/0` shows the data bag is populated as expected. Importantly the credentials field is as follows:

```
credentials: '{"05445a7c9caedbff38042e8cf56611bd": "secret://77577413-d57f-481b-854f-a1231e017724/cqcmqbc6qn7vjti34kg0"}'
```

After running `juju attach-resource <my-app> resource=...` to debug some issues on my app, the pod is restarted and I start to get an error `secret not found`. Running `juju show-unit <my-app>/0` shows that the credentials have changed.

```
credentials: '{"05445a7c9caedbff38042e8cf56611bd": "secret:cqcmqbc6qn7vjti34kg0"}'
```

Notice that the model reference is gone so the charm expects to find the secret in the model and fails. I'm not certain whether this is a charm bug or a Juju bug.

### To Reproduce

1. Bootstrap a controller on lxd
2. Add microk8s as a cloud to the controller (this may require some tweaks to microk8s client config depending on your setup as it took me some trial and error to figure out how to get working - [this Matrix discussion](https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$tgvEBnhyruAplcmDMDsFhe1PXuLWpZY25z0FTwVX7Zk?via=ubuntu.com&via=matrix.org&via=mx.aouss.it) may be helpful)
3. `juju add-model source localhost`
4. `juju deploy vault --channel=1.15/stable` 
5. `juju offer vault`
6. `juju add-model destination microk8s`
7. `juju deploy juju-jimm-k8s jimm --channel=latest/edg` (or use any charm that relates to Vault)
8. Relate vault to app - `juju relate jimm <offer-url>`
9. Wait until `juju show-unit jimm/0` shows the data bag is populated.
10. Either run what I did `juju attach-resource jimm jimm-image=foo` or try `kubectl delete pod` on the workload.
11. Run `juju show-unit jimm/0` to see if the data bag is updated as expected.

### Environment

This was consistently replicated locally as described above.
```
$ microk8s version
MicroK8s v1.28.10 revision 6829
```
```
snap info microk8s
tracking:     1.28-strict/stable
```
```
juju version
3.5.2-genericlinux-amd64
```
```
App         Version  Status  Scale  Charm       Channel      Rev  Exposed  Message
vault                active      1  vault       1.15/stable  353  yes
```

### Relevant log output

```shell
unit-jimm-0: 21:37:08 ERROR unit.jimm/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/model.py", line 3049, in _run
    result = subprocess.run(args, **kwargs)  # type: ignore
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-jimm-0/secret-get', 'secret:cqcmqbc6qn7vjti34kg0', '--format=json')' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/model.py", line 3397, in secret_get
    result = self._run('secret-get', *args, return_output=True, use_json=True)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/model.py", line 3051, in _run
    raise ModelError(e.stderr) from e
ops.model.ModelError: ERROR secret "secret:cqcmqbc6qn7vjti34kg0" not found


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-jimm-0/charm/./src/charm.py", line 838, in <module>
    main(JimmOperatorCharm)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/main.py", line 544, in main
    manager.run()
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/main.py", line 520, in run
    self._emit()
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/main.py", line 509, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/main.py", line 143, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/framework.py", line 350, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/framework.py", line 849, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/framework.py", line 939, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-jimm-0/charm/./src/charm.py", line 227, in _on_jimm_pebble_ready
    self._update_workload(event)
  File "/var/lib/juju/agents/unit-jimm-0/charm/src/state.py", line 25, in wrapper
    return func(self, event)
  File "/var/lib/juju/agents/unit-jimm-0/charm/./src/charm.py", line 355, in _update_workload
    vault_config = self._vault_config()
  File "/var/lib/juju/agents/unit-jimm-0/charm/./src/charm.py", line 269, in _vault_config
    secret = self.model.get_secret(id=unit_credentials)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/model.py", line 284, in get_secret
    content = self._backend.secret_get(id=id, label=label)
  File "/var/lib/juju/agents/unit-jimm-0/charm/venv/ops/model.py", line 3400, in secret_get
    raise SecretNotFoundError() from e
ops.model.SecretNotFoundError
```

**Comment by gruyaume** [comment]
Hello @kian99 thank you for opening this issue. We should be looking at it in the next couple of days. Thank you.

**Comment by saltiyazan** [comment]
Issue is being addressed in #254 and a bug report has been created in Juju https://bugs.launchpad.net/juju/+bug/2075153
